tests/cloudpickle_test.py

```diff
@@ -1257,6 +1258,44 @@ def test_dataclass(self):
         pickle_depickle(DataClass, protocol=self.protocol)
         assert data.x == pickle_depickle(data, protocol=self.protocol).x == 42
 
+    def test_dynamically_defined_enum(self):
+
+        class Color(Enum):
+            """3-element color space"""
+            RED = 1
+            GREEN = 2
+            BLUE = 3
+
+        green1, green2, ClonedColor = pickle_depickle(
+            [Color.GREEN, Color.GREEN, Color], protocol=self.protocol)
+        assert green1 is green2
+        assert green1 is ClonedColor.GREEN
+        assert green1 is not ClonedColor.BLUE
+
+        assert ClonedColor.__doc__ == Color.__doc__
+        assert ClonedColor.__module__ == Color.__module__
+
+        if hasattr(Color, "__qualname__"):
+            assert ClonedColor.__qualname__ == Color.__qualname__
+
+        # cloudpickle systematically creates new copies for locally defined
+        # classes that cannot be imported by name:
+        assert green1 is not Color.GREEN
+        assert ClonedColor is not Color
```

Comment:
While the above is technically expected and respects the usual cloudpickle semantics, I believe this will be even more surprising than just "breaking isinstance" (as reported in #195): testing physical identity is actually the expected way to use enums.

In particular, this would mean that the following code that uses interactively defined Enums on a cluster would not work:

```python
from enum import Enum
from dask.distributed import Client

class Color(Enum):
    RED = 1
    GREEN = 2

def check_data(data):
    return Color.GREEN if sum(data) > 0 else Color.RED

client = Client()
client.submit(check_data, [1, 2, 3]).result() is Color.GREEN
```

Therefore I am not sure if supporting interactively defined enums is a good feature for cloudpickle or not...

Reply:
Shouldn't "interactively defined enums" be fetched from `__main__` rather than recreated?

Reply:
Maybe and this should probably be the case for any interactively defined types in `__main__` (see #195) but we should be careful to still make sure that, if later, that implementation changes in the main process, no previous copies will lie around in the `__main__` of worker processes, otherwise the new implementation will never get propagated to the worker process:


```
from enum import Enum
from dask.distributed import Client

class Color(Enum):
    RED = 1
    GREEN = 2

def check_data(data):
    return Color.GREEN if sum(data) >= 0 else Color.RED

client = Client()
client.submit(check_data, [1, 2, 3]).result() is Color.GREEN

class Color(Enum):
    RED = 1
    BLUE = 2

def check_data(data):
    return Color.BLUE if sum(data) >= 0 else Color.RED

client.submit(check_data, [1, 2, 3]).result() is Color.BLUE
```

We currently lack a missing test harness to add non-regression test for those semantics in the cloudpickle test suite. The `subprocess_pickle_echo` thingy does not allow to check for the typical interactive remote call use-case of pyspark / loky / dask-distributed / ray justifies the existence of cloudpickle in the first place. I think improving the test harness in that respect is a priority.

Reply:
I run some interactive tests with loky and the current cloudpickle master, and nothing is leaked (`setattr`'ed on sys.modules["__main__"]) so this should be fine.

Note that locally defined enums as tested in the test of this PR would still be non-lookup-able but I think it would be fine to special case the `__main__` module.

Reply:
I pushed a much better solution that tracks any dynamic class definition and matched unpickled instances back to them.